fix: bootstrap check with alternate anchor properties

kspearrin opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Bug fix: microsoft edge does not support the origin property on anchor elements, so bootstrapping fails
for edge extensions. this fix simply changes the bootstrap origin check to use supported
anchor properties while having the same effect


**What is the current behavior? (You can also link to an open issue here)**
closes #16030 


**What is the new behavior (if this is a feature change)?**
n/a


**Does this PR introduce a breaking change?**
No


**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

kspearrin committed:
fix: bootstrap check with alternate anchor properties

microsoft edge does not support the origin property on anchor elements, so bootstrapping fails
for edge extensions. this fix simply changes the bootstrap origin check to use supported
anchor properties while having the same effect

closes #16030

frederikprijck commented on src/Angular.js:
```diff
@@ -1517,7 +1517,8 @@ function allowAutoBootstrap(document) {
     link.href = src.value;
 
-    if (document.location.origin === link.origin) {
+    if (document.location.protocol === link.protocol && document.location.host === link.host &&
```
```
if (document.location.protocol === link.protocol && document.location.host === link.host &&
      document.location.port === link.port) {
      // Same-origin resources are always allowed, even for non-whitelisted schemes.
      return true;
    }
```

Doesn't checking the `host` already include comparing the `port`? Maybe you where looking for `hostname` (https://developer.mozilla.org/en-US/docs/Web/API/Location)?

What about keep checking `origin`, but provide a fallback (polyfill?) when origin doesn't exist ?
```
if (!window.location.origin) {
  window.location.origin = window.location.protocol + "//" + window.location.hostname + (window.location.port ? ':' + window.location.port: '');
}
```

kspearrin replied:
Ah yes, I got `host` and `hostname` confused. Will update.

You could polyfill it, but I don't know what differences there may be across browsers and how they construct origin with `protocol` + `hostname` + `post` so this seems the best way to compare IMO.

kspearrin committed:
fix: bootstrap check with hostname rather than host

`host` already includes port so be more explicit with `hostname`

closes #16030

kspearrin committed:
fix: bootstrap check with alternate anchor properties tests

tests were outdated and still constructing fake doc with only a protocol and origin property

closes #16030

kspearrin commented:
Sorry about the multiple commits. New to the codebase and just figuring things out :)

Let me know if this is all ok now and if any changes are needed.

Narretz commented:
Two things that come to mind:

- `origin` is a read only property, the other props are not and could be tampered with. Auto-bootstrap in extensions is security relevant, so this might be a problem
- this is a bug in Edge, and I would prefer to get it fixed there.
- the problem can be easily worked around by bootstrapping manually

kspearrin commented:
Good points. I didn't think about these not being read only. I will also file an issue with the edge fracker about the problem.

kspearrin commented:
Issue also opened with the Edge team here: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/12236493/

gkalpak commented:
The concern about other properties being writeable is not valid, since `.origin` reflects the values of the other (writeable) properties anyway, so  it is possible to change origin as well. We are safe here though (famous last words), since we are creating a new anchor element and not exposing it to anyone, so we know its properties haven't been tampered with.

I agree with @Narretz's other two points though:
- This is a bug in Edge, not AngularJS.
- There is already a work-around available (manual bootstrap), so I am not keen on changing the "security-sensitive" `allowAutoBootstrap()` function.

Narretz commented:
@kspearrin Thanks for opening the bug report. It has been assigned already! Hopefully it won't take ages to get it fixed. In the meantime, I'll close this PR, since the workaround is very easy and the frequency is very low - but I'll leave the issue open.
